compiler-core/src/package_interface.rs

```diff
@@ -0,0 +1,280 @@
+                                .iter()
+                                .map(|arg| TypeInterface::from_type(arg.type_.as_ref()))
+                                .collect(),
+                            return_: Box::new(TypeInterface::from_type(return_type)),
+                        },
+                    },
+                );
+            }
+
+            // Private definitions are not included.
+            Definition::Function(_) => {}
+            Definition::CustomType(_) => {}
+            Definition::ModuleConstant(_) => {}
+
+            // Type aliases are ignored since those are also erased from the
+            // generated documentation.
+            Definition::TypeAlias(_) => {}
+
+            // Imports are ignored.
+            Definition::Import(_) => {}
+        }
+    }
+    (types, values)
+}
+
+impl TypeInterface {
+    fn from_type(type_: &Type) -> TypeInterface {
+        match type_ {
+            Type::Fn { args, retrn } => TypeInterface::Fn {
+                parameters: args
+                    .iter()
+                    .map(|arg| TypeInterface::from_type(arg.as_ref()))
+                    .collect(),
+                return_: Box::new(TypeInterface::from_type(retrn)),
+            },
+
+            Type::Tuple { elems } => TypeInterface::Tuple {
+                elements: elems
+                    .iter()
+                    .map(|elem| TypeInterface::from_type(elem.as_ref()))
+                    .collect(),
+            },
+
+            Type::Var { type_ } => match type_.as_ref().try_borrow() {
+                // Package serialisation happens after inference has taken place
+                // At this point no one should be helding a mutable reference of
+                // the type: the borrow should never fail.
+                Err(_) => unreachable!(),
+                Ok(ref_) => match ref_.deref() {
+                    // Since package serialisation happens after inference there
+                    // should be no unbound type variables.
+                    // TODO: This branch should be `unreachable!()` but because of
+                    //       https://github.com/gleam-lang/gleam/issues/2533
+                    //       we sometimes end up with those in top level
+                    //       definitions.
+                    //       However, `Unbound` and `Generic` ids are generated
+                    //       using the same generator so we have no problem treating
+                    //       unbound variables as generic ones since ids will never
+                    //       overlap.
+                    //       Once #2533 is closed this branch can be turned back to
+                    //       be unreachable!().
+                    TypeVar::Unbound { id } => TypeInterface::Variable { id: *id },
+                    TypeVar::Link { type_ } => TypeInterface::from_type(type_),
+                    TypeVar::Generic { id } => TypeInterface::Variable { id: *id },
```

Comment:
This should use new ints for each value, not the internal ints from the compiler. So if a function's type has 3 variables they will always be 0, 1, and 2.